### PR TITLE
Implement APIs for both platform configuration + guest launch ioctls

### PR DESCRIPTION
**Pull request by tfanelli-rh**
Tested on Milan hardware.

### CODE REVIEW

**Comment by haraldh** [comment]
Seems like this will require to bump the minimal rust version to 1.46 in https://github.com/enarx/snp/blob/main/.github/workflows/test.yml#L24

```
error[E0658]: `if` is not allowed in a `const fn`
Error:   --> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/vmm-sys-util-0.7.0/src/linux/epoll.rs:33:1
   |
33 | / bitflags! {
34 | |     /// The type of events we can monitor a file descriptor for.
35 | |     pub struct EventSet: u32 {
36 | |         /// The associated file descriptor is available for read operations.
...  |
68 | |     }
69 | | }
   | |_^
   |
   = note: for more information, see https://github.com/rust-lang/rust/issues/49146
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

**Comment by haraldh** [comment]
For the failing [check-spdx-headers](https://github.com/enarx/snp/pull/7/checks?check_run_id=3318559825#step:3:8) to pass, you just have to git rebase to current origin.

File: src/launch/mod.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        const MIGRATE_MA =  0b0000000000000100;
+
+        /// If enabled, debugging is allowed.
+        const DEBUG =       0b0000000000001000;
+    }
+}
+
+impl PolicyFlags {
+    /// Represent policy flags to their u16 counterpart.
+    pub fn to_u16(&self) -> u16 {
```
Comment (haraldh) [inline src/launch/mod.rs:35]: Why not use `PolicyFlags::bits()` here?

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/launch/mod.rs (hunk 1 above)
Comment (tfanelli-rh) [reply src/launch/mod.rs]: Fixed.

**Comment by tfanelli-rh** [comment]
For some reason, I'm getting panics/errors on the `Doc-tests`.

```
   Doc-tests snp
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/thread/mod.rs:1269:48
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0 (53cb7b09b 2021-06-17) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no --crate-type lib

query stack during panic:
end of query stack
```

**Comment by tfanelli-rh** [comment]
At the moment, this does not use the `sev-iocuddle` crate. I intend to change that with upcoming changes now that I see what details `snp` and `sev` share.

File: src/util/mod.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+use std::slice::{from_raw_parts, from_raw_parts_mut};
+
+pub trait TypeLoad: Read {
+    fn load<T: Sized + Copy>(&mut self) -> Result<T> {
+        #[allow(clippy::uninit_assumed_init)]
+        let mut t = unsafe { MaybeUninit::uninit().assume_init() };
+        let p = &mut t as *mut T as *mut u8;
+        let s = unsafe { from_raw_parts_mut(p, size_of::<T>()) };
+        self.read_exact(s)?;
+        Ok(t)
```
[the remark this answers is not in the document]
Comment (tfanelli-rh) [reply src/util/mod.rs]: Fixed

File: src/launch/mod.rs (hunk 1 above)
Comment (tfanelli-rh) [reply src/launch/mod.rs]: Fixed

**Comment by haraldh** [comment]
> For the failing [check-spdx-headers](https://github.com/enarx/snp/pull/7/checks?check_run_id=3318559825#step:3:8) to pass, you just have to git rebase to current `origin/main`.

This still applies.. please rebase to `origin/main`:
```
$ git fetch origin
$ git rebase origin/main
```

**Comment by haraldh** [comment]
> Seems like this will require to bump the minimal rust version to 1.46 in https://github.com/enarx/snp/blob/main/.github/workflows/test.yml#L24
> 
> ```
> error[E0658]: `if` is not allowed in a `const fn`
> Error:   --> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/vmm-sys-util-0.7.0/src/linux/epoll.rs:33:1
>    |
> 33 | / bitflags! {
> 34 | |     /// The type of events we can monitor a file descriptor for.
> 35 | |     pub struct EventSet: u32 {
> 36 | |         /// The associated file descriptor is available for read operations.
> ...  |
> 68 | |     }
> 69 | | }
>    | |_^
>    |
>    = note: for more information, see https://github.com/rust-lang/rust/issues/49146
>    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
> ```

```diff
diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 7540f52..9f255d6 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -21,7 +21,7 @@ jobs:
           - nightly
           - beta
           - stable
-          - 1.42.0
+          - 1.46.0
         profile:
           - name: debug
           - name: release
```

File: src/launch/mod.rs [hunk 3]
```diff
@@ -139,3 +141,23 @@ impl PageType {
...
 }
+
+/// Encapsulates the data needed to complete a guest launch.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Deserialize, Serialize)]
+pub struct Finish {
+    /// sPA of the ID block. Ignored if ID_BLOCK_EN is 0.
+    pub id_block_uaddr: u64,
+
+    /// sPA of the authentication information of the ID block. Ignored if ID_BLOCK_EN is 0.
+    pub id_auth_uaddr: u64,
```
Comment (connorkuehl) [inline src/launch/mod.rs:152]: Not sPAs.

**Comment by tfanelli-rh** [comment]
> > For the failing [check-spdx-headers](https://github.com/enarx/snp/pull/7/checks?check_run_id=3318559825#step:3:8) to pass, you just have to git rebase to current `origin/main`.
> 
> This still applies.. please rebase to `origin/main`:
> 
> ```
> $ git fetch origin
> $ git rebase origin/main
> ```

I've rebased with `origin/main`, but the test still fails.

```
$ git branch
* guest_launch
  main
$ git fetch origin 
$ git rebase origin/main
Current branch guest_launch is up to date.
$ git push -f origin guest_launch 
Everything up-to-date

```

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Nice job, @tfanelli-rh!

There are a couple of things that need to change for this PR.

The first, I suspect, is that you've been reading too closely to the SNP firmware spec. It's vital that we understand the firmware spec, but it isn't _technically_ the interface this code should be programming against. The actual interface you need to program against is the Linux kernel ioctls exported to userspace by Brijesh's hypervisor series.

The reason why I mention this at all is because your doc comments associated with each of the structs mention system physical addresses as well as taking only certain bits of them. Brijesh's ioctl documentation doesn't mention these at all, and as expected, they seem to only care about userspace virtual addresses. 

Please double check the doc comments you're exporting as well as the fields to make sure they align with the structs exported by the kernel. Please update your doc comments accordingly.

The second thing is that I think we should also pause on this PR, get `sev-iocuddle` package-worthy, release it to crates.io, and then finally consume it here in `snp`. Similarly, the `sev` crate will need to be updated to be downstream of `sev-iocuddle` so we can realize the benefits of factoring that code out into `sev-iocuddle` in the first place. Otherwise, we run the risk of maintaining the same code in multiple places, which is not fun.

**Comment by connorkuehl** [comment]
> > > For the failing [check-spdx-headers](https://github.com/enarx/snp/pull/7/checks?check_run_id=3318559825#step:3:8) to pass, you just have to git rebase to current `origin/main`.
> > 
> > 
> > This still applies.. please rebase to `origin/main`:
> > ```
> > $ git fetch origin
> > $ git rebase origin/main
> > ```
> 
> I've rebased with `origin/main`, but the test still fails.
> 
> ```
> $ git branch
> * guest_launch
>   main
> $ git fetch origin 
> $ git rebase origin/main
> Current branch guest_launch is up to date.
> $ git push -f origin guest_launch 
> Everything up-to-date
> ```

What is your `origin` remote?

```
$ git remote get-url origin
```

If it's not this repo, you'll want to either:

```
$ git remote add upstream https://github.com/enarx/snp
$ git fetch upstream
$ git rebase upstream/master
```

or just a one-off (not recommended):

```
$ git fetch https://github.com/enarx/snp master
$ git rebase FETCH_HEAD
```

**Comment by tfanelli-rh** [comment]
Regarding the `check-spdx-headers` test failing:
```
$ git fetch upstream
$ git rebase upstream/main
Current branch guest_launch is up to date.
```
I'll take a look into why this is.

**Comment by connorkuehl** [comment]
Could it be a bug in the spdx action?

```
2021-08-20T14:50:33.9757504Z Traceback (most recent call last):
2021-08-20T14:50:33.9759198Z   File "/home/runner/work/_actions/enarx/spdx/master/verify-spdx-headers", line 120, in <module>
2021-08-20T14:50:33.9760158Z     print(f"{license:16} {path}")
2021-08-20T14:50:33.9761015Z TypeError: unsupported format string passed to NoneType.__format__
```

**Comment by tfanelli-rh** [comment]
> Regarding the `check-spdx-headers` test failing:
> 
> ```
> $ git fetch upstream
> $ git rebase upstream/main
> Current branch guest_launch is up to date.
> ```
> 
> I'll take a look into why this is.

Fixed.

**Comment by haraldh** [comment]
> > Seems like this will require to bump the minimal rust version to 1.46 in https://github.com/enarx/snp/blob/main/.github/workflows/test.yml#L24
> > ```
> > error[E0658]: `if` is not allowed in a `const fn`
> > Error:   --> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/vmm-sys-util-0.7.0/src/linux/epoll.rs:33:1
> >    |
> > 33 | / bitflags! {
> > 34 | |     /// The type of events we can monitor a file descriptor for.
> > 35 | |     pub struct EventSet: u32 {
> > 36 | |         /// The associated file descriptor is available for read operations.
> > ...  |
> > 68 | |     }
> > 69 | | }
> >    | |_^
> >    |
> >    = note: for more information, see https://github.com/rust-lang/rust/issues/49146
> >    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
> > ```
> 
> ```diff
> diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
> index 7540f52..9f255d6 100644
> --- a/.github/workflows/test.yml
> +++ b/.github/workflows/test.yml
> @@ -21,7 +21,7 @@ jobs:
>            - nightly
>            - beta
>            - stable
> -          - 1.42.0
> +          - 1.46.0
>          profile:
>            - name: debug
>            - name: release
> ```

Just add a commit bumping the version and you'll be fine

**Review by haraldh — APPROVED** [review]

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Blocking until it can consume sev-iocuddle from crates.io.

**Pull request closed without merging**